### PR TITLE
Continual Transformer Encoder

### PR DESCRIPTION
**Pull request by LukasHedegaard**
New tool for efficient application of Transformer Encoders to time-series data.
Note: this does not come with pre-trained models. The user must fit the learner on her own features.

### CODE REVIEW

**Comment by ad-daniel** [comment]
I assume this PR is still ongoing, I've converted it to draft for now

**Comment by LukasHedegaard** [comment]
> I assume this PR is still ongoing, I've converted it to draft for now

It is ready for review

**Review by ad-daniel — CHANGES_REQUESTED** [review]
Thank you, only a few points
- ensure the default values in the code are indeed the desired ones and adapt the docs/comments accordingly
- add a changelog entry mentioning the added tool
- if possible, provide a demo that shows how to use it (like you did in the docs, can be the same one)
- Concerning the project structure, my understanding is that it's a self-contained module. You've made (rightly so) an ad hoc folder for this new tool in perception/activity_recognition and test folders, but documentation wise you've put everything in the same existing document. I think it would be better to split the documentation of x3d and this tool into 2 separate documents. Basically as we did for other stuff like speech recognition, and each document should be referenced in the index here : https://github.com/opendr-eu/opendr/blob/master/docs/reference/index.md
- I don't think this was among the planned tools, so there isn't a strict requirement, but having a ROS/ROS2 node would be a plus. Up to you.

**Comment by LukasHedegaard** [comment]
Thank you for updating the docs and adjusting the formatting, @ad-daniel .

Per your suggestion, I've updated the structure of the docs, splitting documentation for this module into its own file and updating index.md accordingly. Here, I've also collapsed the "action recognition" and "activity recognition" categories into one (action/activity is often used interchangeably in the literature).

Moreover, the reason for the delay to this PR was a large augmentation to the `continual-inference` library that enables exporting ONNX models with the accelerated "forward_step" mode. This module has been updated with that augmentation, which is enabled given the correct library version (torch >= 1.10, onnxrunetime >= 1.11, and continual-inference >= 1.0.0). Dynamic checks are made at runtime to enforce these.

**Comment by LukasHedegaard** [comment]
There seem to be a lot of failures in tools other than this one since develop was merged (perception/activity_recognition passes). I assume this is not related to this tool, but let me know if can be of any help.

**Comment by ad-daniel** [comment]
I've re-run the test a couple of times and it fails rather consistently (although in other PR targeting develop everything passes nearly every time), so I assume that since the dependency constraints here are a bit less strict, some dependencies are installed differently. One of the failures is due to this https://github.com/opendr-eu/opendr/pull/369, and as some stuff isn't installed due to the error it leads to problems. Might be worth waiting until that PR is merged

**Comment by ad-daniel** [comment]
No problem, I believe the demo and changelog entry are still missing though, no? (I assume you won't be providing a ROS interface at the moment?)

**Comment by LukasHedegaard** [comment]
> 

My bad. I've updated the CHANGELOG and added a demo. I will not be providing a ROS node at this time.

**Review by ad-daniel — CHANGES_REQUESTED** [review]
Sorry, last detail, could you add also a README.md file to the demos folder with some instructions? other than that it's fine for me

**Review by ad-daniel — DISMISSED** [review]
Thank you!

**Comment by LukasHedegaard** [comment]
> Thank you!

Likewise :-)

File: projects/python/perception/activity_recognition/demos/continual_transformer_encoder/README.md
Comment (tsampazk) [inline projects/python/perception/activity_recognition/demos/continual_transformer_encoder/README.md:1]:
```suggestion
# Continual Transformer Encoder demo
```

File: projects/python/perception/activity_recognition/demos/continual_transformer_encoder/README.md
Comment (tsampazk) [inline projects/python/perception/activity_recognition/demos/continual_transformer_encoder/README.md:3]:
```suggestion
The file [demo.py](demo.py) is a demo of how to use the `CoTransEncLearner`, including fitting, evaluation, runtime optimization and inference.
```

File: src/opendr/perception/activity_recognition/continual_transformer_encoder/continual_transformer_encoder_learner.py
Comment (tsampazk) [inline src/opendr/perception/activity_recognition/continual_transformer_encoder/continual_transformer_encoder_learner.py:46]: Is this class actually used?

File: docs/reference/continual-transformer-encoder.md
Comment (tsampazk) [inline docs/reference/continual-transformer-encoder.md:110]: There's a full stop missing here.
```suggestion
This method is used to perform classification of a video.
Returns a `engine.target.Category` objects, where each holds a category.
```

File: docs/reference/continual-transformer-encoder.md
Comment (tsampazk) [inline docs/reference/continual-transformer-encoder.md:151]:
```suggestion
Optimize model execution. This is accomplished by saving to the ONNX format and loading the optimized model.
```

File: src/opendr/perception/activity_recognition/continual_transformer_encoder/continual_transformer_encoder_learner.py
Comment (tsampazk) [inline src/opendr/perception/activity_recognition/continual_transformer_encoder/continual_transformer_encoder_learner.py:577]:
```suggestion
        This is accomplished by saving to the ONNX format and loading the optimized model.
```

**Review by tsampazk — CHANGES_REQUESTED** [review]
Thank you Lukas! I left some comments regarding some minor typos, etc.

File: src/opendr/perception/activity_recognition/continual_transformer_encoder/continual_transformer_encoder_learner.py
Comment (LukasHedegaard) [reply src/opendr/perception/activity_recognition/continual_transformer_encoder/continual_transformer_encoder_learner.py]: No it's not. Not sure how I missed that one 😅 I've removed it in a follow-up commit. Thanks.

File: src/opendr/perception/activity_recognition/continual_transformer_encoder/continual_transformer_encoder_learner.py
Comment (LukasHedegaard) [reply src/opendr/perception/activity_recognition/continual_transformer_encoder/continual_transformer_encoder_learner.py:507]: Good catch. Names were changed to "x" in both cases. (see commit a14dca89050d64).

File: docs/reference/continual-transformer-encoder.md
Comment (LukasHedegaard) [reply docs/reference/continual-transformer-encoder.md]: Good catch. Names were changed to "x" in both cases. (see commit a14dca89050d64).

File: src/opendr/perception/activity_recognition/datasets/dummy_timeseries_dataset.py
Comment (LukasHedegaard) [reply src/opendr/perception/activity_recognition/datasets/dummy_timeseries_dataset.py]: Indeed. Fixed in 092bb003

File: src/opendr/perception/activity_recognition/README.md
Comment (LukasHedegaard) [reply src/opendr/perception/activity_recognition/README.md:67]: Fixed in 62c6bce96

**Review by tsampazk — APPROVED** [review]
Thank you!

**Review by ad-daniel — APPROVED** [review]
Thank you!

**Pull request merged**
